eclair-core/src/main/scala/fr/acinq/eclair/blockchain/bitcoind/ZmqWatcher.scala

```diff
@@ -239,13 +251,134 @@ class ZmqWatcher(chainHash: ByteVector32, blockCount: AtomicLong, client: Extend
+            Future.successful(fundTxResponse.copy(tx = txSingleOutput))
+          case None =>
+            client.getChangeAddress()(singleThreadExecutionContext).map(pubkeyHash => {
+              val txSingleOutput = fundTxResponse.tx.copy(txOut = Seq(TxOut(dummyChangeAmount, Script.pay2wpkh(pubkeyHash))))
+              fundTxResponse.copy(tx = txSingleOutput)
+            })
+        }
+      }).map(fundTxResponse => {
+        require(fundTxResponse.tx.txOut.size == 1, "funded transaction should have a single change output")
+        // NB: we insert the anchor input in the *first* position because our signing helpers only sign input #0.
+        val unsignedTx = txWithInput.copy(tx = fundTxResponse.tx.copy(txIn = txWithInput.tx.txIn.head +: fundTxResponse.tx.txIn))
+        adjustAnchorOutputChange(unsignedTx, commitTx, fundTxResponse.amountIn + Transactions.AnchorOutputsCommitmentFormat.anchorAmount, currentFeerate, targetFeerate, dustLimit)
+      }).flatMap(claimAnchorTx => {
+        val claimAnchorSig = keyManager.sign(claimAnchorTx, localFundingPubKey, Transactions.TxOwner.Local, commitmentFormat)
+        val signedClaimAnchorTx = Transactions.addSigs(claimAnchorTx, claimAnchorSig)
+        val commitInfo = ExtendedBitcoinClient.PreviousTx(signedClaimAnchorTx.input, signedClaimAnchorTx.tx.txIn.head.witness)
+        client.signTransaction(signedClaimAnchorTx.tx, Seq(commitInfo))(singleThreadExecutionContext)
+      }).flatMap(signTxResponse => {
+        client.publishTransaction(signTxResponse.tx)(singleThreadExecutionContext)
+      })
+    }
+  }
+
+  def publishHtlcTx(currentFeerate: FeeratePerKw, targetFeerate: FeeratePerKw, dustLimit: Satoshi, signingKit: TransactionSigningKit.HtlcTxSigningKit): Future[ByteVector32] = {
+    import signingKit._
+    if (targetFeerate <= currentFeerate) {
+      val localSig = keyManager.sign(txWithInput, localHtlcBasepoint, localPerCommitmentPoint, Transactions.TxOwner.Local, commitmentFormat)
+      val signedHtlcTx = addHtlcTxSigs(txWithInput, localSig, signingKit)
+      log.info("publishing htlc tx without adding inputs: txid={}", signedHtlcTx.tx.txid)
+      client.publishTransaction(signedHtlcTx.tx)(singleThreadExecutionContext)
+    } else {
+      log.info("publishing htlc tx with additional inputs: commit input={}:{} target feerate={}", txWithInput.input.outPoint.txid, txWithInput.input.outPoint.index, targetFeerate)
+      val txNotFunded = txWithInput.tx.copy(txIn = Nil, txOut = txWithInput.tx.txOut.head.copy(amount = dustLimit) :: Nil)
+      val htlcTxWeight = signingKit match {
+        case _: TransactionSigningKit.HtlcSuccessSigningKit => commitmentFormat.htlcSuccessWeight
+        case _: TransactionSigningKit.HtlcTimeoutSigningKit => commitmentFormat.htlcTimeoutWeight
+      }
+      // NB: bitcoind will add at least one P2WPKH input.
+      val weightRatio = htlcTxWeight.toDouble / (txNotFunded.weight() + Transactions.claimP2WPKHOutputWeight)
```

Comment:
It's similar to what we did in `publishAnchorTx` where we target a higher feerate because bitcoind doesn't see our non-wallet input. I've been lazy with that one and use an approximation that worked well in practice (it overestimates the feerate) but let me do the maths to find an (almost) exact formula.

Reply:
Done in https://github.com/ACINQ/eclair/pull/1702/commits/378108ab76f2e936b3984c298403d5dc2aba519f

I updated to use the exact formula (not riemann hypothesis math level, but I still had to scratch my head a bit).
The issue with that exact formula is that we don't take into account the fee already paid by the htlc (htlc input - htlc output) and it would be a real pain to take it fully into account...
I think it's fine for now, the only drawback is that if bitcoind doesn't add any change output we won't be able to offset this, so we'll end up paying a bigger fee (which is why I had to change the test to allow a bigger error margin). We could add a change output to correct that (with an additional call to bitcoind) but I think it's not worth it, in practice this is going to be a really low amount.

Reply:
Nice!